File: c-sharp/RiskSystem/SerialPricer.cs
using System.Reflection;
using HmxLabs.TechTest.Models;

namespace HmxLabs.TechTest.RiskSystem
{
    public class SerialPricer : PricerBase
    {
        public void Price(IEnumerable<IEnumerable<ITrade>> tradeContainters, IScalarResultReceiver resultReceiver, string assemblyPath)
        {
            LoadPricers(assemblyPath);

            foreach (var tradeContainter in tradeContainters)
            {
                foreach (var trade in tradeContainter)
                {
                    if (!TryGetPricer(trade.TradeType, out var pricer))
                    {
                        resultReceiver.AddError(trade.TradeId, "No Pricing Engines available for this trade type");
                        continue;
                    }

                    pricer.Price(trade, resultReceiver);
                }
            }
        }
    }
}

File: cpp/RiskSystem/SerialPricer.cpp
#include "SerialPricer.h"

void SerialPricer::loadPricers() {
    if (!pricers_.empty()) {
        return;
    }

    pricers_ = PricingEngineFactory::createPricersFromConfigFile("./PricingConfig/PricingEngines.xml");
}

void SerialPricer::priceTrade(ITrade& trade, IScalarResultReceiver& resultReceiver) {
    const std::string tradeType = trade.getTradeType();
    if (pricers_.find(tradeType) == pricers_.end()) {
        resultReceiver.addError(trade.getTradeId(), "No Pricing Engines available for this trade type");
        return;
    }

    pricers_.at(tradeType)->price(&trade, &resultReceiver);
}

void SerialPricer::price(ITrade& trade, IScalarResultReceiver& resultReceiver) {
    loadPricers();
    priceTrade(trade, resultReceiver);
}

void SerialPricer::price(const std::vector<std::vector<std::unique_ptr<ITrade>>>& tradeContainers,
                         IScalarResultReceiver& resultReceiver) {
    loadPricers();

    for (const auto& tradeContainer : tradeContainers) {
        for (const auto& trade : tradeContainer) {
            priceTrade(*trade, resultReceiver);
        }
    }
}
